File: Toggly.FeatureManagement.Node/toggly-node-core/src/telemetry/metrics-batcher.ts
import { toProtobufTimestamp } from './grpc-clients.js'

export interface MetricsFeatureOptions {
  /** Optional feature key for experiment-style subcounts. */
  feature?: string
  /** Variant name (default: enabled). */
  variant?: string
}

export interface MetricsBatcherOptions {
  appKey: string
  environment: string
  instanceName?: string
}

type MetricKey = string // `${metric}\0${feature ?? ''}`

export interface MetricStatPayload {
  appKey: string
  environment: string
  time: { seconds: number; nanos: number }
  stats: Array<{
    metric: string
    feature?: string
    variantValues: Record<string, number>
  }>
  counters: Array<{
    metric: string
    feature?: string
    variantValues: Record<string, number>
  }>
  observations: Array<{
    time: { seconds: number; nanos: number }
    metric: string
    feature?: string
    variantValues: Record<string, number>
  }>
  instanceName?: string
}

/**
 * In-memory business metrics aggregator using variantValues maps (.NET parity).
 */
export class MetricsBatcher {
  private readonly appKey: string
  private readonly environment: string
  private readonly instanceName?: string
  private measures = new Map<MetricKey, Map<string, number>>()
  private counters = new Map<MetricKey, Map<string, number>>()
  private observations: Array<{
    time: Date
    metric: string
    feature?: string
    variant: string
    value: number
  }> = []

  constructor(options: MetricsBatcherOptions) {
    this.appKey = options.appKey
    this.environment = options.environment
    this.instanceName = options.instanceName
  }

  private key(metric: string, feature?: string): MetricKey {
    return `${metric}\0${feature ?? ''}`
  }

  private parseKey(key: MetricKey): { metric: string; feature?: string } {
    const [metric, feature = ''] = key.split('\0')
    return feature ? { metric, feature } : { metric }
  }

  private addToMap(
    store: Map<MetricKey, Map<string, number>>,
    metric: string,
    value: number,
    options?: MetricsFeatureOptions,
  ): void {
    const variant = options?.variant ?? 'enabled'
    const mapKey = this.key(metric, options?.feature)
    let variants = store.get(mapKey)
    if (!variants) {
      variants = new Map()
      store.set(mapKey, variants)
    }
    variants.set(variant, (variants.get(variant) ?? 0) + value)
  }

  measure(metric: string, value: number, options?: MetricsFeatureOptions): void {
    this.addToMap(this.measures, metric, value, options)
  }

  incrementCounter(metric: string, value = 1, options?: MetricsFeatureOptions): void {
    this.addToMap(this.counters, metric, value, options)
  }

  observe(metric: string, value: number, options?: MetricsFeatureOptions): void {
    this.observations.push({
      time: new Date(),
      metric,
      feature: options?.feature,
      variant: options?.variant ?? 'enabled',
      value,
    })
  }

  isEmpty(): boolean {
    return this.measures.size === 0 && this.counters.size === 0 && this.observations.length === 0
  }

  private drainMap(
    store: Map<MetricKey, Map<string, number>>,
  ): Array<{ metric: string; feature?: string; variantValues: Record<string, number> }> {
    const out: Array<{ metric: string; feature?: string; variantValues: Record<string, number> }> =
      []
    for (const [key, variants] of store) {
      const parsed = this.parseKey(key)
      const variantValues: Record<string, number> = {}
      for (const [variant, value] of variants) {
        if (value !== 0) {
          variantValues[variant] = value
        }
      }
      if (Object.keys(variantValues).length > 0) {
        out.push({
          metric: parsed.metric,
          ...(parsed.feature ? { feature: parsed.feature } : {}),
          variantValues,
        })
      }
    }
    store.clear()
    return out
  }

  buildAndReset(): MetricStatPayload | null {
    if (this.isEmpty()) {
      return null
    }

    const observationGroups = new Map<
      string,
      {
        time: { seconds: number; nanos: number }
        metric: string
        feature?: string
        variantValues: Record<string, number>
      }
    >()

    for (const obs of this.observations) {
      const groupKey = `${obs.time.toISOString()}\0${obs.metric}\0${obs.feature ?? ''}`
      let group = observationGroups.get(groupKey)
      if (!group) {
        group = {
          time: toProtobufTimestamp(obs.time),
          metric: obs.metric,
          ...(obs.feature ? { feature: obs.feature } : {}),
          variantValues: {},
        }
        observationGroups.set(groupKey, group)
      }
      group.variantValues[obs.variant] = obs.value
    }
    this.observations = []

    const payload: MetricStatPayload = {
      appKey: this.appKey,
      environment: this.environment,
      time: toProtobufTimestamp(),
      stats: this.drainMap(this.measures),
      counters: this.drainMap(this.counters),
      observations: [...observationGroups.values()],
    }

    if (this.instanceName) {
      payload.instanceName = this.instanceName
    }

    return payload
  }
}

File: Toggly.FeatureManagement.Python/toggly/src/toggly/telemetry/metrics_batcher.py
"""In-memory business metrics aggregator (Metrics.SendMetrics payload shape)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, List, Mapping, Optional, Tuple

from toggly.telemetry.grpc_clients import to_protobuf_timestamp

MetricKey = Tuple[str, str]  # (metric, feature or "")


@dataclass
class MetricsFeatureOptions:
    """Optional feature/variant correlation for a metric sample."""

    feature: Optional[str] = None
    variant: Optional[str] = None


class MetricsBatcher:
    """Accumulate business metrics using ``variantValues`` maps (.NET parity)."""

    def __init__(
        self,
        app_key: str,
        environment: str,
        *,
        instance_name: Optional[str] = None,
    ) -> None:
        self._app_key = app_key
        self._environment = environment
        self._instance_name = instance_name
        self._measures: Dict[MetricKey, Dict[str, float]] = {}
        self._counters: Dict[MetricKey, Dict[str, float]] = {}
        self._observations: List[Dict[str, Any]] = []

    @staticmethod
    def _key(metric: str, feature: Optional[str]) -> MetricKey:
        return (metric, feature or "")

    def _add_to_map(
        self,
        store: Dict[MetricKey, Dict[str, float]],
        metric: str,
        value: float,
        options: Optional[MetricsFeatureOptions] = None,
    ) -> None:
        variant = "enabled"
        feature: Optional[str] = None
        if options is not None:
            if options.variant:
                variant = options.variant
            feature = options.feature
        map_key = self._key(metric, feature)
        variants = store.get(map_key)
        if variants is None:
            variants = {}
            store[map_key] = variants
        variants[variant] = variants.get(variant, 0.0) + value

    def measure(
        self,
        metric: str,
        value: float,
        options: Optional[MetricsFeatureOptions] = None,
    ) -> None:
        """Aggregate a measurement (sum over the flush window)."""
        self._add_to_map(self._measures, metric, value, options)

    def increment_counter(
        self,
        metric: str,
        value: float = 1.0,
        options: Optional[MetricsFeatureOptions] = None,
    ) -> None:
        """Increment a counter."""
        self._add_to_map(self._counters, metric, value, options)

    def observe(
        self,
        metric: str,
        value: float,
        options: Optional[MetricsFeatureOptions] = None,
    ) -> None:
        """Record a point-in-time observation (gauge)."""
        variant = "enabled"
        feature: Optional[str] = None
        if options is not None:
            if options.variant:
                variant = options.variant
            feature = options.feature
        self._observations.append(
            {
                "time": datetime.now(timezone.utc),
                "metric": metric,
                "feature": feature,
                "variant": variant,
                "value": value,
            }
        )

    def is_empty(self) -> bool:
        return not self._measures and not self._counters and not self._observations

    def _drain_map(
        self, store: Dict[MetricKey, Dict[str, float]]
    ) -> List[Dict[str, Any]]:
        out: List[Dict[str, Any]] = []
        for (metric, feature), variants in store.items():
            variant_values = {
                name: value for name, value in variants.items() if value != 0
            }
            if not variant_values:
                continue
            item: Dict[str, Any] = {"metric": metric, "variantValues": variant_values}
            if feature:
                item["feature"] = feature
            out.append(item)
        store.clear()
        return out

    def build_and_reset(self) -> Optional[Dict[str, Any]]:
        """Build a MetricStat-shaped dict and clear aggregates."""
        if self.is_empty():
            return None

        observation_groups: Dict[str, Dict[str, Any]] = {}
        for obs in self._observations:
            when: datetime = obs["time"]
            group_key = f"{when.isoformat()}\0{obs['metric']}\0{obs.get('feature') or ''}"
            group = observation_groups.get(group_key)
            if group is None:
                group = {
                    "time": to_protobuf_timestamp(when),
                    "metric": obs["metric"],
                    "variantValues": {},
                }
                if obs.get("feature"):
                    group["feature"] = obs["feature"]
                observation_groups[group_key] = group
            group["variantValues"][obs["variant"]] = obs["value"]
        self._observations = []

        payload: Dict[str, Any] = {
            "appKey": self._app_key,
            "environment": self._environment,
            "time": to_protobuf_timestamp(),
            "stats": self._drain_map(self._measures),
            "counters": self._drain_map(self._counters),
            "observations": list(observation_groups.values()),
        }
        if self._instance_name:
            payload["instanceName"] = self._instance_name
        return payload


def metrics_options(
    feature: Optional[str] = None,
    variant: Optional[str] = None,
) -> MetricsFeatureOptions:
    """Convenience constructor for :class:`MetricsFeatureOptions`."""
    return MetricsFeatureOptions(feature=feature, variant=variant)


def options_from_mapping(
    options: Optional[Mapping[str, Any]],
) -> Optional[MetricsFeatureOptions]:
    """Accept dict-style options from public client APIs."""
    if options is None:
        return None
    return MetricsFeatureOptions(
        feature=options.get("feature"),
        variant=options.get("variant"),
    )
